jena-arq/src/main/java/org/apache/jena/riot/resultset/rw/ResultSetWriterXML.java

```diff
@@ -253,84 +254,138 @@ public void binding(String varName, RDFNode node) {
-                printLiteral((Literal)node);
+            Node node = rdfNode.asNode();
+            printBindingValue(node);
+        }
+        
+        private void printBindingValue(Node node) {
+            if ( node == null )
+                return;
+        
+            if ( node.isLiteral() ) {
+                printLiteral(node);
                 return;
             }
 
-            if ( node instanceof Resource ) {
-                printResource((Resource)node);
+            if ( node.isURI() ) {
+                printURI(node);
+                return;
+            }
+            
+            if ( node.isBlank() ) {
+                printBlankNode(node);
+                return;
+            }
+            if ( node.isNodeTriple() ) {
+                printTripleTerm(node);
                 return;
             }
+            
+            if ( node.isNodeGraph() )
+                throw new UnsupportedOperationException("Graph terms");
 
-            Log.warn(this, "Unknown RDFNode type in result set: " + node.getClass());
+            Log.warn(this, "Unknown RDFNode type in result set: " + node);
         }
 
-        void printLiteral(Literal literal) {
+        private void printURI(Node nodeURI) {
+            String uri = nodeURI.getURI();
             out.print("<");
-            out.print(dfLiteral);
+            out.print(dfURI);
+            out.print(">");
+            out.print(xml_escape(uri));
+            out.print("</");
+            out.print(dfURI);
+            out.println(">");
+        }            
+            
+        private void printBlankNode(Node node) {
+            String label = bNodeMap.get(null, node);
+            // Comes with leading "_:"
+            label = label.substring(2);
+            out.print("<");
+            out.print(dfBNode);
+            out.print(">");
+            out.print(xml_escape(label));
+            out.print("</");
+            out.print(dfBNode);
+            out.println(">");
+        }
 
+        private void printLiteral(Node literal) {
+            out.print("<");
+            out.print(dfLiteral);
+        
             if ( Util.isLangString(literal) ) {
-                String lang = literal.getLanguage();
+                String lang = literal.getLiteralLanguage();
                 out.print(" xml:lang=\"");
-                out.print(literal.getLanguage());
+                out.print(lang);
                 out.print("\"");
             } else if ( !Util.isSimpleString(literal) ) {
                 // Datatype
                 // (RDF 1.1) not xsd:string nor rdf:langString.
                 // (RDF 1.0) any datatype.
-                String datatype = literal.getDatatypeURI();
+                String datatype = literal.getLiteralDatatypeURI();
                 out.print(" ");
                 out.print(dfAttrDatatype);
                 out.print("=\"");
                 out.print(datatype);
                 out.print("\"");
             }
-
+        
```

Comment:
Extra white space

Reply:
I didn't remove all trailing WS because it is all over the place.